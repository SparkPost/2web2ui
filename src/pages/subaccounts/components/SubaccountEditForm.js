import React, { Component } from 'react';
import { connect } from 'react-redux';
import { reduxForm, formValueSelector } from 'redux-form';
import { tokens } from '@sparkpost/design-tokens-hibana';
import { Box, Button, Panel, Stack } from 'src/components/matchbox';
import { ButtonWrapper } from 'src/components';
import { Form } from 'src/components/form';
import { getIpPools } from 'src/selectors/ipPools';
import { selectFirstIpPoolId } from 'src/selectors/ipPools';
import { NameField, StatusSelect } from './formFields';
import IpPoolSelect from './IpPoolSelect';
import RestrictToIpPoolCheckbox from './RestrictToIpPoolCheckbox';

export class SubaccountEditForm extends Component {
  render() {
    const {
      handleSubmit,
      pristine,
      submitting,
      ipPools,
      compliance,
      reset,
      restrictedToIpPool,
    } = this.props;

    return (
      <Form onSubmit={handleSubmit} id="subaccount-edit-form">
        <Panel.LEGACY.Section>
          <Stack>
            <Box maxWidth={tokens.sizing_1200}>
              <NameField disabled={submitting || compliance} />
            </Box>

            <Box maxWidth={tokens.sizing_1200}>
              <StatusSelect disabled={submitting || compliance} compliance={compliance} />
            </Box>

            {Boolean(ipPools.length) && (
              <Stack>
                <RestrictToIpPoolCheckbox disabled={submitting || compliance} />
                {restrictedToIpPool && (
                  <IpPoolSelect disabled={submitting || compliance} options={ipPools} />
                )}
              </Stack>
            )}
          </Stack>
        </Panel.LEGACY.Section>
        <Panel.LEGACY.Section>
          <ButtonWrapper>
            <Button variant="primary" submit disabled={pristine || submitting || compliance}>
              {submitting ? 'Updating...' : 'Update Subaccount'}
            </Button>
            <Button variant="secondary" disabled={pristine || submitting} onClick={reset}>
              Cancel
            </Button>
          </ButtonWrapper>
        </Panel.LEGACY.Section>
      </Form>
    );
  }
}

const formName = 'SubaccountEditForm';
const valueSelector = formValueSelector(formName);

const mapStateToProps = (state, { subaccount }) => {
  const { compliance, ip_pool, name, status } = subaccount;

  return {
    compliance,
    ipPools: getIpPools(state),
    restrictedToIpPool: valueSelector(state, 'restrictedToIpPool'),
    initialValues: {
      ipPool: ip_pool || selectFirstIpPoolId(state),
      name,
      restrictedToIpPool: Boolean(ip_pool),
      status: compliance ? `${status} by SparkPost` : status,
    },
  };
};

const SubaccountEditReduxForm = reduxForm({ form: formName })(SubaccountEditForm);
export default connect(mapStateToProps, {})(SubaccountEditReduxForm);
