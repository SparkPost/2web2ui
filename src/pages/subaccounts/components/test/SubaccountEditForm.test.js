import { shallow } from 'enzyme';
import React from 'react';

import { SubaccountEditForm } from '../SubaccountEditForm';

const handleSubmit = jest.fn();

const props = {
  submitting: false,
  pristine: true,
  compliance: false,
  ipPools: ['an array of ip pools'],
  handleSubmit
};

let wrapper;

beforeEach(() => {
  wrapper = shallow(<SubaccountEditForm {...props} />);
});

test('base case', () => {
  expect(wrapper).toMatchSnapshot();
});

test('no ip pools', () => {
  wrapper.setProps({ ipPools: []});
  expect(wrapper).toMatchSnapshot();
});

test('with an assigned ip pool', () => {
  wrapper.setProps({ ipPool: 'test', restrictedToIpPool: true });
  expect(wrapper).toMatchSnapshot();
});

test('cancel button', () => {
  wrapper.setProps({ pristine: false });
  expect(wrapper).toMatchSnapshot();
});

test('submits correctly', () => {
  wrapper.find('Form').simulate('submit');
  expect(handleSubmit).toHaveBeenCalled();
  wrapper.setProps({ submitting: true });
  expect(wrapper).toMatchSnapshot();
});

test('should disable all fields when compliance is true', () => {
  wrapper.setProps({ compliance: true });
  expect(wrapper).toMatchSnapshot();
});
