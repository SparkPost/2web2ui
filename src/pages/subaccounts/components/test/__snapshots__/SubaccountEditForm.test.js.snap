// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base case 1`] = `
<Form
  id="subaccount-edit-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <Box
        maxWidth="45rem"
      >
        <NameField
          disabled={false}
        />
      </Box>
      <Box
        maxWidth="45rem"
      >
        <StatusSelect
          compliance={false}
          disabled={false}
        />
      </Box>
      <Stack>
        <RestrictToIpPoolCheckbox
          disabled={false}
        />
      </Stack>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        disabled={true}
        submit={true}
        variant="primary"
      >
        Update Subaccount
      </Button>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`cancel button 1`] = `
<Form
  id="subaccount-edit-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <Box
        maxWidth="45rem"
      >
        <NameField
          disabled={false}
        />
      </Box>
      <Box
        maxWidth="45rem"
      >
        <StatusSelect
          compliance={false}
          disabled={false}
        />
      </Box>
      <Stack>
        <RestrictToIpPoolCheckbox
          disabled={false}
        />
      </Stack>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        disabled={false}
        submit={true}
        variant="primary"
      >
        Update Subaccount
      </Button>
      <Button
        disabled={false}
        variant="secondary"
      >
        Cancel
      </Button>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`no ip pools 1`] = `
<Form
  id="subaccount-edit-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <Box
        maxWidth="45rem"
      >
        <NameField
          disabled={false}
        />
      </Box>
      <Box
        maxWidth="45rem"
      >
        <StatusSelect
          compliance={false}
          disabled={false}
        />
      </Box>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        disabled={true}
        submit={true}
        variant="primary"
      >
        Update Subaccount
      </Button>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`should disable all fields when compliance is true 1`] = `
<Form
  id="subaccount-edit-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <Box
        maxWidth="45rem"
      >
        <NameField
          disabled={true}
        />
      </Box>
      <Box
        maxWidth="45rem"
      >
        <StatusSelect
          compliance={true}
          disabled={true}
        />
      </Box>
      <Stack>
        <RestrictToIpPoolCheckbox
          disabled={true}
        />
      </Stack>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        disabled={true}
        submit={true}
        variant="primary"
      >
        Update Subaccount
      </Button>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`submits correctly 1`] = `
<Form
  id="subaccount-edit-form"
  onSubmit={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": undefined,
    }
  }
>
  <Panel.LEGACY.Section>
    <Stack>
      <Box
        maxWidth="45rem"
      >
        <NameField
          disabled={true}
        />
      </Box>
      <Box
        maxWidth="45rem"
      >
        <StatusSelect
          compliance={false}
          disabled={true}
        />
      </Box>
      <Stack>
        <RestrictToIpPoolCheckbox
          disabled={true}
        />
      </Stack>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        disabled={true}
        submit={true}
        variant="primary"
      >
        Updating...
      </Button>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`with an assigned ip pool 1`] = `
<Form
  id="subaccount-edit-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <Box
        maxWidth="45rem"
      >
        <NameField
          disabled={false}
        />
      </Box>
      <Box
        maxWidth="45rem"
      >
        <StatusSelect
          compliance={false}
          disabled={false}
        />
      </Box>
      <Stack>
        <RestrictToIpPoolCheckbox
          disabled={false}
        />
        <IpPoolSelect
          disabled={false}
          options={
            Array [
              "an array of ip pools",
            ]
          }
        />
      </Stack>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        disabled={true}
        submit={true}
        variant="primary"
      >
        Update Subaccount
      </Button>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;
